File: code/socket2/python/echo_EPETserv.py
import socket
import select
import sys
import errno

BUF_SIZE = 4
EPOLL_SIZE = 50


def set_nonblocking(sock):
    """将 socket 设为非阻塞模式"""
    sock.setblocking(False)


def main():
    if len(sys.argv) != 2:
        print(f"Usage : {sys.argv[0]} <port>")
        sys.exit(1)

    serv_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    serv_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    serv_sock.bind(("", int(sys.argv[1])))
    serv_sock.listen(5)

    epfd = select.epoll()
    fd_to_sock = {serv_sock.fileno(): serv_sock}

    # 监听 socket 设为非阻塞
    set_nonblocking(serv_sock)
    epfd.register(serv_sock.fileno(), select.EPOLLIN)

    try:
        while True:
            events = epfd.poll(EPOLL_SIZE, -1)
            if not events:
                continue

            print("return epoll_wait")
            for fd, event in events:
                if fd == serv_sock.fileno():
                    # 有新客户端连接
                    clnt_sock, clnt_adr = serv_sock.accept()
                    set_nonblocking(clnt_sock)
                    # 注意：注册时加上 EPOLLET（边缘触发）
                    epfd.register(clnt_sock.fileno(),
                                  select.EPOLLIN | select.EPOLLET)
                    fd_to_sock[clnt_sock.fileno()] = clnt_sock
                    print(f"connected client: {clnt_sock.fileno()}")
                else:
                    clnt_sock = fd_to_sock.get(fd)
                    if clnt_sock is None:
                        continue

                    # 边缘触发模式下，必须一直读直到 EAGAIN
                    while True:
                        try:
                            data = clnt_sock.recv(BUF_SIZE)
                        except BlockingIOError:
                            # 对应 C 里的 errno == EAGAIN，数据读完了
                            break
                        except ConnectionResetError:
                            data = b""

                        if not data:
                            # str_len == 0，客户端关闭
                            epfd.unregister(fd)
                            clnt_sock.close()
                            del fd_to_sock[fd]
                            print(f"closed client: {fd}")
                            break
                        else:
                            # echo 回去
                            clnt_sock.sendall(data)
    except KeyboardInterrupt:
        pass
    finally:
        serv_sock.close()
        epfd.close()


if __name__ == "__main__":
    main()

File: code/socket2/cxx/echo_EPETserv.cpp
#include "socket.hpp"

#include <cstdlib>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace unpsock;

int main(int argc, char* argv[])
{
    if (argc != 2) {
        std::cerr << "Usage : " << argv[0] << " <port>\n";
		exit(1);
    }

    // ---- 监听 socket ----
    Socket serv_sock(Family::INET, AddrType::STREAM);
    serv_sock.set_reuse_addr(true);

    auto serv_addr = Address::any(std::atoi(argv[1]), Family::INET);
    serv_sock.bind(serv_addr);
    serv_sock.listen(5);
    serv_sock.set_nonblocking(true);   // 原程序对监听 socket 也非阻塞

    EpollSelector sel;
    sel.init();
    // 监听 socket：LT（不带 EdgeTrigger），与原程序一致
    sel.add(serv_sock, Event::Read);

    std::unordered_map<int, Socket> clients;

    const int BUF_SIZE = 4;
    while (true) {
        auto ready = sel.wait();   // 永久阻塞
        std::cout << "return epoll_wait" << std::endl;

        for (const auto& ev : ready) {
            int fd = ev.fd;

            // 异常 / 挂断
            if (has_event(ev.events, Event::Error) ||
                    has_event(ev.events, Event::Hangup)) {
                if (fd != serv_sock.fileno()) {
                    sel.remove(fd);
                    clients.erase(fd);
                    std::cout << "closed client: " << fd << std::endl;
                }
                continue;
            }

            if (!has_event(ev.events, Event::Read)) continue;

            if (fd == serv_sock.fileno()) {
                // ---- 新连接（监听 socket 用 LT，accept 一次即可）----
                auto [conn, peer] = serv_sock.accept();
                int conn_fd = conn.fileno();

                // 客户端：非阻塞 + 边缘触发
                conn.set_nonblocking(true);
                sel.add(conn, Event::Read | Event::EdgeTrigger);   // ← 就这里不同

                clients.emplace(conn_fd, std::move(conn));
                std::cout << "connected client: " << conn_fd << std::endl;
            } else {
                // ---- 客户端数据：ET 必须循环读到 EAGAIN ----
                auto it = clients.find(fd);
                if (it == clients.end()) continue;

                Socket& conn = it->second;

                while (true) {
                    std::error_code ec;
                    auto data = conn.recv(BUF_SIZE, ec);

                    if (ec) {
                        if (ec == std::errc::resource_unavailable_try_again) {
                        } else {
                            std::cout << "recv fd=" << fd << ": " << ec.message() << std::endl;
                        }
                        break;
                    }
                    if (data.empty()) {              // 对端关闭
                        // 对端关闭
                        sel.remove(fd);
                        clients.erase(it);
                        std::cout << "closed client: " << fd << std::endl;
                        break;
                    }

                    conn.sendall(data);
                }
            }
        }
    }

    return 0;
}
